barretenberg/cpp/src/barretenberg/vm2/tracegen/execution_trace.test.cpp

```diff
@@ -234,5 +310,182 @@ TEST(ExecutionTraceGenTest, Gas)
                       Contains(Field(&R::execution_limit_used_da_cmp_diff, 16000))));
 }
 
+TEST(ExecutionTraceGenTest, SimpleDiscardSingleEvent)
+{
+    TestTraceContainer trace;
+    ExecutionTraceBuilder builder;
+
+    // Just test a single event to isolate the issue
+    std::vector<simulation::ExecutionEvent> events;
+
+    // Single ADD operation
+    events.push_back(createAddEvent(1, 0, TransactionPhase::APP_LOGIC));
+
+    builder.process(events, trace);
+
+    const auto& rows = trace.as_rows();
+    EXPECT_GE(rows.size(), 1);
+}
+
+TEST(ExecutionTraceGenTest, DiscardNestedRevertContext)
+{
+    TestTraceContainer trace;
+    ExecutionTraceBuilder builder;
+
+    // Create a sequence: parent context calls child context, child does some work then reverts
+    std::vector<simulation::ExecutionEvent> events;
+
+    // Event 1: Parent context does ADD
+    events.push_back(createAddEvent(1, 0, TransactionPhase::APP_LOGIC));
+
+    // Event 2: Parent calls child (context 1 -> 2)
+    events.push_back(createCallEvent(1, 0, TransactionPhase::APP_LOGIC, 2));
+
+    // Event 3: Child context does ADD - this should have discard=1 since child will revert
+    events.push_back(createAddEvent(2, 1, TransactionPhase::APP_LOGIC));
+
+    // Event 4: Child context reverts
+    events.push_back(createErrorEvent(2, 1, TransactionPhase::APP_LOGIC, 1));
+
+    // Event 5: Parent continues after child reverted
+    events.push_back(createAddEvent(1, 0, TransactionPhase::APP_LOGIC));
+
+    // Event 6: Parent returns successfully (top-level exit)
+    events.push_back(createReturnEvent(1, 0, TransactionPhase::APP_LOGIC));
+
+    builder.process(events, trace);
+
+    const auto& rows = trace.as_rows();
```

Comment:
this should be either `const auto` or `auto` but not a reference. `as_rows()` returns an object. This is only working because of https://stackoverflow.com/questions/39718268/why-do-const-references-extend-the-lifetime-of-rvalues